### PR TITLE
 Index entity conditionally

### LINKED ISSUE #202: Conditionally index entities
It would be nice to be able to conditionally index entities based on specific criteria (like a field returning true or false) like the only lib allowed for.

### PR DESCRIPTION
**Pull request by julienbourdeau**
Fix #202 

Add new `index_if` configuration key for indices.
This should be the name of a method in the entity that returns true if the item should be index (false to bypass indexing).

### CODE REVIEW

File: src/IndexManager.php [hunk 1]
```diff
@@ -157,6 +163,17 @@ public function count($query, $className, array $parameters = [])
         return $this->engine->count($query, $this->getFullIndexName($className));
     }
 
+    protected function shouldBeIndexed($className, &$entity)
+    {
+        if ($methodName = $this->indexIfMapping[$className]) {
+            if (method_exists($entity, $methodName)) {
```
Comment (tristanbes) [inline src/IndexManager.php]: why don't you use `PropertyAccess` ?  https://symfony.com/doc/current/components/property_access.html

It's more flexible, can read from Object and Arrays and is safe

Comment (julienbourdeau) [reply src/IndexManager.php]: Why not but it depends what for. Did you mean to replace to `method_exists` call or do you see a way to rewrite the function in a more elegant way?

Comment (tristanbes) [reply src/IndexManager.php]:
```php
$propertyAccessor = PropertyAccess::createPropertyAccessor();

if ($propertyAccessor->isReadable($entity, $accessor) {
    return (bool) $propertyAccessor->getValue($entity, $accessor)
}
```

Also, with this, you should rename `methodName`  to `accessor` for better clarity in my opinion.

Comment (julienbourdeau) [reply src/IndexManager.php]: In this case, I'm not sure what value this brings. Is there a case where `method_exists` will fail but the PropertyAccess will work?

Comment (tristanbes) [reply src/IndexManager.php]: `method_exist` will return false in this case: 

```yml
  index_if: indexable
```

```php
protected $indexable = false;
// no getter or setter for $indexable.
```

when `$propertyAccessor->getValue($entity, 'indexable'))` does work

Also this won't work with `method_exists`:

```yml
  index_if: config.indexable
```

```php
protected $config = ['indexable' => false];
```

when `$propertyAccessor->getValue($entity, 'config.indexable'))` does work

--- 
well you're in Symfony context, using property accessor will bring more flexibility, I really don't see any disavantage of using a component of the ecosystem you're developing a bundle for.

Comment (julienbourdeau) [reply src/IndexManager.php]: I didn't think about class properties, in this case, it does bring a lot a value! 🚀 I'll update the PR. Thanks for reviewing :)

Comment (tristanbes) [reply src/IndexManager.php]: 👍

**Comment by arteniioleg** [comment]
> `false` to bypass indexing

... and to remove from index on update.

Scenario:

1. Create post `isPublished = false` (not index)
2. Update post `isPublished = true` (index)
3. Update post `isPublished = false` (remove from index)

**Comment by julienbourdeau** [comment]
@arteniioleg Yes that would be great, I'll have a look to see if it can be done quickly.

**Comment by julienbourdeau** [comment]
@arteniioleg I added this feature in the Doctrine event listener, I think it's where it makes the most sense. If I add it to `IndexManager::index`, it will had a lot of operation. Also, it's a bit weird if `index` actually behave like `remove`.

Please let me know what you think!

**Comment by arteniioleg** [comment]
@julienbourdeau Agree :+1:

File: src/Command/SearchImportCommand.php [hunk 2]
```diff
@@ -46,7 +46,7 @@ protected function execute(InputInterface $input, OutputInterface $output)
 
                 $output->writeln(sprintf(
```
Comment (aequasi) [inline src/Command/SearchImportCommand.php]: A Progressbar would be cool for these too :)

**Comment by aequasi** [comment]
It makes sense in an event listener, but as of now, there's no easy way to bulk-reindex/remove the entire index atm (from what I can see).

So, if I mark 50 or so entities as `isPublished = false` for example, in the DB, the import command won't remove them from algolia

File: src/Command/SearchImportCommand.php (hunk 2 above)
Comment (julienbourdeau) [reply src/Command/SearchImportCommand.php]: That would be awesome 🥇

**Comment by julienbourdeau** [comment]
Currently, the only way to affect changes like this is to remove them explicitely or reindex entirely.
I realize there is not one mention of `remove` in the documentation 😅 

To reindex, you can run both command `php bin/console search:clear --indices=posts` and `php bin/console search:import --indices=posts`.

To remove them, you can do something like that after your SQL call:

```php
$idsToRemove = ...; // from SQL

$em = $this->getDoctrine()->getManagerForClass(Post::class);
$repo = $em->getRepository(Post::class);
$entites = $repo->findBy(['id' => $idsToRemove]);

$indexManager->remove($entities, $em)
```

There might even be a way to trigger the doctrine event of doctrine, I'm not sure.

**Pull request merged**